crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+//! Implementing a plugin is done in two steps:
+//!
+//! 1. Create a struct that implements `PluginBuilder`
+//!     - Its purpose is to simply instantiate your plugins as needed with custom logic if required
+//! 2. Create your plugin struct that implements `Plugin`
+
+use futures::future::BoxFuture;
+use std::{
+    any::{Any, TypeId},
+    collections::HashSet,
+};
+
+use downcast_rs::{impl_downcast, DowncastSync};
+
+use async_trait::async_trait;
+
+use crate::{
+    address::{InternalMessage, ReceiverBundle, ReplySender},
+    error::{PluginError, DirectoryError},
+    message::CoreMessages,
+    Address,
+};
+
+/// The communication struct to interface with the core of ThinEdge
+///
+/// Implementors of this trait can be used to get an address of a certain plugin, which can then be
+/// used to send messages of a specific type to that plugin.
+/// Alternatively, implementors of this trait can be used to send messages to the core of
+/// thin-edge.
+///
+/// # Note
+///
+/// As a plugin author, you will not have to implement this trait.
+/// The core of thin-edge will use this trait to hand over an object to a plugin that can then be
+/// used to communicate with other plugins (as described above).
+///
+pub trait PluginDirectory: Send + Sync {
```

Comment:
When is a plugin instance added to the underlying directory?

Since, there is no `register` method, I guess that this is done behind the seen. But when? Can I use an address directly in the `instanciate` function creating a plugin?

Reply:
Technically, you can send messages as early as instantiate, since internally it uses a channel. But the receiver end is not yet listening. This means that after N messages it will block! (N is set by the configuration and shared between _all_ Addr to that plugin)